[자동차 경주 구현] 빙티 미션 제출합니다.

s6m1n opened the pull request:
### 자동차 경주 Step2
Step1에서 리뷰해주신 덕에 궁금했던 점들을 해소할 수 있었습니다 감사합니다 😄
남겨주신 피드백들 모두 적용했으며, 이번 단계에 대한 질문은 아래와 같습니다. Step2도 잘 부탁드려요!

페어와 미션을 진행할 당시에는 `OutputView`에서 Model인 `Car`에 직접 접근하는 것을 막고자,
`Controller`에서 `Car` 내부 프로퍼티인 `name`과 `position`을 꺼내어 전달했습니다.
그러나 추후 View에서는 Model에 의존해도 된다는 것을 알고 `printCurrentPosition`에서 List<Car>를 인자로 받아 처리하도록 수정했습니다.
이렇게 List로 받으니, 해당 메소드 안에서 List가 끝나는 시점을 알고 개행 문자를 출력할 수 있어
이전의 '컨트롤러에서 println()을 수행해도 될까?' 에 대한 고민이 사라졌습니다.

대신 새로운 고민이 생겼습니다! 🤔
`Car` 클래스에서 toString을 아래처럼 override하여 구현해볼까 했는데, 이 부분은 View의 책임이라고 느껴져 적용하지 않았습니다.
`CURRENT_POSITION_TEMPLATE.format(this.name, PROGRESS_BAR.repeat(this.position))`
**꼭 출력을 수행하지 않더라도, 출력을 위해 형식을 맞춰주는 과정도 View의 책임으로 볼 수 있을까요?**
그렇다면 '각 자동차의 진행 상황을 전진 횟수 1회 당 - 기호 한 개로 표시한다'라는 뷰 로직이라고 봐야하는 걸까요?

해당 미션과는 별개이지만, 서버로부터 오는 데이터의 형식과 화면에 보여줄 형식이 달라 로직이 꽤 복잡한 mapper를 만들었던 경험이 있습니다.
이런 경우 코드를 어느 layer에 위치시켜야 할지 고민이었는데, 리뷰어님은 비즈니스 로직과 뷰 로직을 어떤 기준으로 정해주고 있는지에 대해 여쭤보고 싶습니다!

s6m1n committed:
docs: 기능 목록 작성

s6m1n committed:
test: 자동차 이름 입력에 대한 테스트케이스 추가

Co-authored-by: kmkim2689 <101035437+kmkim2689@users.noreply.github.com>

s6m1n committed:
test: 시도 횟수 입력 테스트 케이스 추가

Co-authored-by: kmkim2689 <101035437+kmkim2689@users.noreply.github.com>

s6m1n committed:
chore: 입력값에 대한 테스트 클래스명 변경

Co-authored-by: kmkim2689 <101035437+kmkim2689@users.noreply.github.com>

s6m1n committed:
chore: 자동차 이름 입력 예외 메시지 상수 추가

Co-authored-by: kmkim2689 <101035437+kmkim2689@users.noreply.github.com>

s6m1n committed:
feat: 자동차 이름 입력 유효성 검사 메서드 구현

Co-authored-by: kmkim2689 <101035437+kmkim2689@users.noreply.github.com>

s6m1n committed:
docs: 기능 목록 업데이트

Co-authored-by: kmkim2689 <101035437+kmkim2689@users.noreply.github.com>

s6m1n committed:
feat: 필요한 유효 정보들을 사용자로부터 입력 받는 기능 추가

Co-authored-by: kmkim2689 <101035437+kmkim2689@users.noreply.github.com>

s6m1n committed:
feat: 랜덤 메서드를 이용한 이동 여부 결정 기능 구현

Co-authored-by: kmkim2689 <101035437+kmkim2689@users.noreply.github.com>

s6m1n committed:
feat: Car 객체 생성 후 이동 로직 구현

Co-authored-by: kmkim2689 <101035437+kmkim2689@users.noreply.github.com>

s6m1n committed:
feat: 게임에 필요한 메시지 출력 메소드 추가

Co-authored-by: kmkim2689 <101035437+kmkim2689@users.noreply.github.com>

s6m1n committed:
feat: 게임 흐름에 따른 메시지 출력 구현

Co-authored-by: kmkim2689 <101035437+kmkim2689@users.noreply.github.com>

s6m1n committed:
fix: 최종 우승자를 출력하는 중복된 메소드 삭제

Co-authored-by: kmkim2689 <101035437+kmkim2689@users.noreply.github.com>

s6m1n committed:
test: printCurrentPosition 메소드에 대한 테스트 케이스 추가

Co-authored-by: kmkim2689 <101035437+kmkim2689@users.noreply.github.com>

s6m1n committed:
refactor: 랜덤 넘버 생성 로직을 MoveManager에서 분리

Co-authored-by: kmkim2689 <101035437+kmkim2689@users.noreply.github.com>

s6m1n committed:
refactor: 랜덤 숫자 추출 관련 상수들의 위치 Constants로 변경

Co-authored-by: kmkim2689 <101035437+kmkim2689@users.noreply.github.com>

s6m1n committed:
chore: InputValidatorTest 파일 패키지 이동

s6m1n committed:
build: ktlint-idea 플러그인 추가

s6m1n committed:
chore : ktlint를 위한 코드 스타일 수정

s6m1n committed:
chore: format을 이용해 메시지에 포함된 매직 넘버 상수화

s6m1n committed:
refactor: private var _position으로 객체 외부에서의 접근 및 변경을 제한

s6m1n committed:
refactor: RacingController를 class로, currentRacingStatus를 List로 변경

s6m1n committed:
refactor: 개행 출력 메소드 OutputView로 분리

s6m1n committed:
refactor: 랜덤 숫자 추출 메서드 코드 리뷰 반영

s6m1n committed:
refactor: printCurrentPosition에서 List<Car>를 인자로 받아 처리하도록 수정

ghojeong commented:
먼저 질문에 대한 답변드립니다.

# 1. View 의 책임

> 꼭 출력을 수행하지 않더라도, 출력을 위해 형식을 맞춰주는 과정도 View의 책임으로 볼 수 있을까요?

Car 클래스의 책임은 무엇일까요? 
Car의 책임은 자동차의 동작(전진)과 상태 입니다.
따라서 Car 는 동작과 상태에 대한 비즈니스 로직을 갖는 것이 맞습니다.
반면 toString 은 Car 내부 상태에 대한 출력, 즉 View 와 더 관련이 있어보입니다. 

안드로이드 앱 개발을 할 때를 생각해보시면 더 이해하기 쉬우실 것 같아요!
자동차의 움직임을 그리는 코드가 Car 안에 있으면, Car 는 안드로이드 화면을 그려주는 "외부 라이브러리" 가 요구하는 
출력 형식을 반드시 알아야 하는 상황이  생기지 않을까요?

그렇게 되면 Car 를 개발하면서 View 쪽 로직과 데이터 구조를 계속 살펴봐야 합니다.
모델은 View 를 모르더라도 개발이 가능하도록 부탁드립니다.

# 2. 비즈니스 로직과 뷰 로직

> 서버로부터 오는 데이터의 형식과 화면에 보여줄 형식이 달라 로직이 꽤 복잡한 mapper를 만들었던 경험이 있습니다.
이런 경우 코드를 어느 layer에 위치시켜야 할지 고민이었는데, 리뷰어님은 비즈니스 로직과 뷰 로직을 어떤 기준으로 정해주고 있는지에 대해 여쭤보고 싶습니다!

빙티님께서 느끼셨던 것처럼 서버에서 받는 데이터 형식과 UI를 그릴때 필요한 데이터 형식은 다를때가 정말 많습니다! 질문하신 내용이 정말 공감이 갑니다.

- 데이터의 형식을 변환하거나 매핑하는 로직은 비즈니스로직과 거리가 멀어보입니다.
- 그렇다고 변환하는 코드를 뷰단에 두자니 뷰와도 관계가 없어보입니다.

아니 그럼 어느 레이어에 속해야하는거지?🤔 저도 동일한 고민을 했었던 기억이 나네요!
관련한 키워드로 레이어와 레이어 사이의 데이터 이동을 뜻하는 "Data Transfer" 라는 키워드로 공부를 해보시기를 추천드립니다.

ghojeong commented on src/main/kotlin/racingcar/controller/RacingController.kt:
```diff
@@ -6,19 +6,21 @@ import racingcar.model.RandomNumberGenerator
-        repeat(trialNum) { play() }
-        printFinalWinners(getWinners())
-    }
+    private lateinit var currentRacingStatus: List<Car>
```
> RacingController를 class로, currentRacingStatus를 List로 변경해 멀티 쓰레드 환경에서 발생할 수 있는 문제를 해결할 수 있을 것 같습니다.

피드백 반영 잘해주셨네요 👍

ghojeong commented on src/main/kotlin/racingcar/constants/Constants.kt:
```diff
@@ -1,21 +1,24 @@
-
-    const val INVALID_NAME_SIZE_EXCEPTION = "이름은 1~5글자 사이로 입력해주세요."
+    const val INVALID_NAME_SIZE_EXCEPTION =
+        "이름은 $MINIMUM_NAME_LENGTH~${MAXIMUM_NAME_LENGTH}글자 사이로 입력해주세요"
     const val DUPLICATED_NAME_EXCEPTION = "서로 중복되지 않는 이름을 입력해주세요."
-    const val INVALID_NUM_OF_CARS_EXCEPTION = "2대 이상의 자동차가 필요합니다."
+    const val INVALID_NUM_OF_CARS_EXCEPTION = "${MINIMUM_CARS_NUMBER}대 이상의 자동차가 필요합니다."
```
유연하게 수정해주셔서 메시지에 관해 수동으로 해야하는 QA 과정을 막아주셨네요! 👍  💯

ghojeong commented on src/main/kotlin/racingcar/model/Car.kt:
```diff
@@ -2,7 +2,12 @@ package racingcar.model
-}
\ No newline at end of file
+    val position: Int
+        get() = _position
```
초기 위치인 `0` 이 매직넘버로 보입니다! 상수화하는건 어떨까요?

그리고 position 에 직접 접근해서, 수정하는일이 생기지 않도록 잘 제한해 주셨습니다. 👍 
kotlin 에서는 이런 방법도 있다고 조언 남겨봅니다.

```kotlin
class Car() {
    var position = 0
    private set
```

s6m1n replied:
position을 클래스 바디 내부로 옮기고 private setter를 적용하면 Car 생성 시 인자에 default가 아닌 값을 전달해 초기화 할 수 없어지는 걸까요? 만약 그렇다면, 해당 부분에 대한 테스트 코드를 수정해야하는데 시간이 촉박해 적용해보지 못했습니다..
private setter를 활용하는 방법을 모르고 있었는데 코멘트 남겨주셔서 감사합니다.
추후 학습을 이어가 다음 미션에 활용해보겠습니다! 😊

ghojeong commented on src/test/kotlin/racingcar/model/RandomNumberGeneratorTest.kt:
```diff
@@ -17,6 +17,6 @@ class RandomNumberGeneratorTest {
     fun `선택된 랜덤 숫자가 0에서 9 사이에 있는지의 여부 테스트`() {
         val randomNumber = randomNumberGenerator.getRandomNumber()
-        assertThat(randomNumber).isBetween(MINIMUM_RANDOM_SELECTION_NUM, MAXIMUM_RANDOM_SELECTION_NUM - 1)
+        assertThat(randomNumber).isBetween(MINIMUM_RANDOM_SELECTION_NUM, MAXIMUM_RANDOM_SELECTION_NUM)
```
> Random.nextInt는 제가 구현하지 않은 부분이기에 테스트 또한 불필요해 보입니다. 

불필요하다고 말씀하신 것처럼 위의 테스트는 삭제해도 괜찮을 것 같습니다!
테스트를 많이 작성해서 쌓이다보면, 테스트를 돌리는 시간을 줄이기 위해, 불필요한 테스트를 삭제하는 유지보수를 하기도 합니다.

ghojeong commented on src/main/kotlin/racingcar/utils/InputValidator.kt:
```diff
@@ -5,28 +5,37 @@ import racingcar.constants.Constants.INVALID_NAME_SIZE_EXCEPTION
+            trialNum
+        } catch (e: NumberFormatException) {
+            throw IllegalArgumentException(INVALID_TRIAL_NUM_TYPE_EXCEPTION)
+        }
```
try catch 가 input type 에 대한 검증과 trialNum 범위에 대한 검증 전체를 감싸고 있어서 읽기가 어려워보입니다.
작성하신 코드를 아래와 같은 방법을 시도해 볼 수 있다고 꿀팁을 남겨봅니다.

##  1. type 검증만 try catch

```kt
val trialNum = try { input.toInt() } catch (e: NumberFormatException) {
    throw IllegalArgumentException(INVALID_TRIAL_NUM_TYPE_EXCEPTION)
}
```

##  2. toIntOrNull() 사용하기

try catch 대신, String 타입의 input 을 Int 로 변환에 실패하면 null을 반환하는 `toIntOrNull()` 를 사용해서 type 변환 실패를 처리할 수도 있습니다!

```kt
val trialNum = input.toIntOrNull() ?: throw IllegalArgumentException(INVALID_TRIAL_NUM_TYPE_EXCEPTION)
```

코틀린에는 이런 문법이 있구나! 하고 소개해 보았습니다. 😄

s6m1n replied:
기존에 검증 전체를 감싸고 있던 해당 부분을 각각 getValidatedTrialNum와 getIntegerOrException로 분리해 함수가 더 작은 일을 처리하도록 수정하였습니다! 테스트 코드도 분리된 로직에 맞게 수정했습니다. 거기에 알려주신 toIntOrNull 까지 적용하니 코드가 훨씬 간결해졌네요,, 감사합니다!😂👍

ghojeong commented on src/main/kotlin/racingcar/view/OutputView.kt:
```diff
@@ -7,18 +7,22 @@ import racingcar.constants.Constants.INPUT_TRIAL_NUM_MESSAGE
 import racingcar.constants.Constants.WINNERS_SEPARATOR
+import racingcar.model.Car
 
 object OutputView {
+
+    fun printNewLine() = println()
```
println()을 OutputView 로 옮기셨군요! 단순한 코드일지라도 역할과 책임에 대해 명시적으로 집고 넘어가려는 모습이 멋집니다 👍

참고로 저는 이런 명시적인 메서드 선언을 좋아하지만,
리뷰어에 따라, 팀 문화에 따라 메서드 선언을 굳이 하지 말자고 하는 경우도 있으니,
상황에 따라 다르게 대처해주시면 됩니다.

ghojeong commented on src/main/kotlin/racingcar/controller/RacingController.kt:
```diff
@@ -6,19 +6,21 @@ import racingcar.model.RandomNumberGenerator
 import racingcar.view.OutputView
 
-object RacingController {
+class RacingController {
     private val manager = MoveManager()
     private val numberGenerator = RandomNumberGenerator()
```
RacingController 에서 MoveManager 와 RandomNumberGenerator 인스턴스를 직접 생성하고있는데요,
실구현체와 RacingController 가 강하게 결합되어있다는 느낌을 받았습니다!
어떻게 결합도를 낮출 수 있을까요?

가령 MoveManager 의 경우 기획팀에서 4 이상의 숫자가 아니라,
전진할 수 있는 기준을 낮추어 2 이상의 숫자에서 전진할 수 있도록 하고,
기존의 4라는 숫자에 비교해 고객반응이 얼마나 좋아졌는지 A/B 테스트를 하자고 하면 어떻게 하시려 하나요?

또한 NumberGenerator 에서 숫자를 순수하게 랜덤하게 생성하는게 아니라,
자동차 경주에 캐시 아이템을 구현해서, 캐시 아이템을 사용한 이후 3번 동안은 항상 숫자 9가 나오도록 하자는 기획이 나오면,
어떻게 대응하려 하시나요?

안드로이드 앱을 개발하다보면 충분히 일어날 수 있는 상황인지라 여쭈어봅니다.

s6m1n replied:
제시해주신 상황이라면 문제가 될 것 같아 결합도를 낮추기 위해 아래와 같은 시도를 해봤습니다!

**`MoveManager` 인터페이스의 구현체인 `MoveManagerWithMinNum`를 구현했습니다.**
- minNum을 밖에서 주입해 전진 기준에 대한 요구사항이 바뀌어도 객체를 재활용 가능
- 최솟값을 기준으로 움직임을 관리한다는 의미를 객체 이름에 반영

**`NumberGenerator` 인터페이스의 구현체인 `RandomNumberGenerator`를 구현했습니다.**
- includeMinNum, includeMaxNum를 밖에서 주입해 요구사항이 바뀌어도 객체를 재활용 가능
- getNumber이란 보편적인 네이밍 사용, 이후 아이템 로직등이 추가됨에 따라 수정 가능

ghojeong reviewed (changes requested):
안녕하세요 빙티님!
지난 Step 에서 말씀드렸던 피드백 매우 훌륭하게 반영해주셨습니다. 👍 
Step2 는 좀더 난이도를 높여서, 앱 개발 도중 일어날 수 있는 상황에,
유연히 대응할 수 있도록 구조를 요구해보려고 합니다.
만약 많이 어려우시다면 언제든 편하게 DM 주세요!

s6m1n committed:
chore: 불필요한 RandomNumberGeneratorTest.kt 삭제

s6m1n committed:
chore: DEFAULT_POSITION 상수 추가

s6m1n committed:
refactor: getValidatedTrialNum과 getIntegerOrException 로직 분리

s6m1n committed:
test: getValidatedTrialNum과 getIntegerOrException 각각에 대한 테스트 구현

s6m1n committed:
refactor: MoveManager 인터페이스 생성

s6m1n committed:
refactor: MoveManagerWithMinNum 클래스 구현
- MOVE_THRESHOLD를 밖에서 주입해 요구사항이 바뀌어도 객체를 재활용 가능
- 최솟값을 기준으로 움직임을 관리한다는 의미를 객체 이름에 반영

s6m1n committed:
test: MoveManagerWithMinNum 테스트 구현

s6m1n committed:
refactor: NumberGenerator 인터페이스 추가

s6m1n committed:
refactor: RandomNumberGenerator에 숫자 추출 범위 설정을 위한 파라미터 추가

s6m1n commented:
새롭게 알게 된 내용이 많아 앞으로 공부할 생각에 설레네요ㅎㅎ
현업에서 일어날 법한 상황을 제시하며 코멘트 남겨주신 부분이 인상깊었습니다.
Step2까지 바쁜 시간 쪼개어 정성껏 리뷰해주셔서 감사합니다!

ghojeong reviewed (approved):
빙티님!
함께 자동차 미션 하면서 즐거웠습니다.
특히 제 피드백을 굉장히 적극적으로 수용해주셔서 감사합니다.
사실 테스트 커버리지는 별로 중요하지 않습니다.
앞으로 미션을 하며 기능들이 점점 어려워질 텐데,
커버리지에 집착하기 보다는 재밌고 유의미한 테스트에 집중하셔서
쭉 테스트에 흥미를 가지셨으면 좋겠습니다.
완주까지 화이팅입니다. 🔥
